<a href="https://colab.research.google.com/github/Affiwhizz/lab-extractive-question-answering/blob/main/lab-extractive-question-answering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# LAB | Extractive Question Answering

This notebook demonstrates how Pinecone helps you build an extractive question-answering application. To build an extractive question-answering system, we need three main components:

- A vector index to store and run semantic search
- A retriever model for embedding context passages
- A reader model to extract answers

We will use the SQuAD dataset, which consists of **questions** and **context** paragraphs containing question **answers**. We generate embeddings for the context passages using the retriever, index them in the vector database, and query with semantic search to retrieve the top k most relevant contexts containing potential answers to our question. We then use the reader model to extract the answers from the returned contexts.

Let's get started by installing the packages needed for notebook to run:

In [1]:
import os
from dotenv import load_dotenv, find_dotenv
_ = load_dotenv(find_dotenv())

OPENAI_API_KEY  = os.getenv('OPENAI_API_KEY')
PINECONE_API_KEY= os.getenv('PINECONE_API_KEY')

In [6]:
import os
from getpass import getpass

if not os.getenv("OPENAI_API_KEY"):
    os.environ["OPENAI_API_KEY"] = getpass("Paste your OpenAI API key here: ")

if not os.getenv("HUGGINGFACEHUB_API_TOKEN"):
    os.environ["HUGGINGFACEHUB_API_TOKEN"] = getpass("Paste your Hugging Face token here: ")

print("Keys successfully loaded (hidden safely).")

Keys successfully loaded (hidden safely).


# Install Dependencies

In [7]:
!pip install -qU datasets pinecone-client sentence-transformers torch

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 506.3/506.3 kB 16.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.8/42.8 MB 22.1 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pylibcudf-cu12 25.6.0 requires pyarrow<20.0.0a0,>=14.0.0; platform_machine == "x86_64", but you have pyarrow 21.0.0 which is incompatible.
cudf-cu12 25.6.0 requires pyarrow<20.0.0a0,>=14.0.0; platform_machine == "x86_64", but you have pyarrow 21.0.0 which is incompatible.


# Load Dataset

Now let's load the SQUAD dataset from the HuggingFace Model Hub. We load the dataset into a pandas dataframe and filter the title, question, and context columns, and we drop any duplicate context passages.

In [30]:
from datasets import load_dataset

# load the squad dataset into a pandas dataframe
df = load_dataset("squad", split="train").to_pandas()

In [31]:
df.head()

,id,title,context,question,answers
0,5733be284776f41900661182,University_of_Notre_Dame,"Architecturally, the school has a Catholic cha...",To whom did the Virgin Mary allegedly appear i...,"{'text': ['Saint Bernadette Soubirous'], 'answ..."
1,5733be284776f4190066117f,University_of_Notre_Dame,"Architecturally, the school has a Catholic cha...",What is in front of the Notre Dame Main Building?,"{'text': ['a copper statue of Christ'], 'answe..."
2,5733be284776f41900661180,University_of_Notre_Dame,"Architecturally, the school has a Catholic cha...",The Basilica of the Sacred heart at Notre Dame...,"{'text': ['the Main Building'], 'answer_start'..."
3,5733be284776f41900661181,University_of_Notre_Dame,"Architecturally, the school has a Catholic cha...",What is the Grotto at Notre Dame?,{'text': ['a Marian place of prayer and reflec...
4,5733be284776f4190066117e,University_of_Notre_Dame,"Architecturally, the school has a Catholic cha...",What sits on top of the Main Building at Notre...,{'text': ['a golden statue of the Virgin Mary'...


In [32]:
# select only title and context column
df = df[["title", "context"]]

# drop rows containing duplicate context passages
df = df.drop_duplicates(subset="context").reset_index(drop=True)

df

,title,context
0,University_of_Notre_Dame,"Architecturally, the school has a Catholic cha..."
1,University_of_Notre_Dame,"As at most other universities, Notre Dame's st..."
2,University_of_Notre_Dame,The university is the major seat of the Congre...
3,University_of_Notre_Dame,The College of Engineering was established in ...
4,University_of_Notre_Dame,All of Notre Dame's undergraduate students are...
...,...,...
18886,Kathmandu,"Institute of Medicine, the central college of ..."
18887,Kathmandu,Football and Cricket are the most popular spor...
18888,Kathmandu,The total length of roads in Nepal is recorded...
18889,Kathmandu,The main international airport serving Kathman...


# Initialize Pinecone Index

The Pinecone index stores vector representations of our context passages which we can retrieve using another vector (query vector). We first need to initialize our connection to Pinecone to create our vector index. For this, we need a free [API key]("https://app.pinecone.io/"), and then we initialize the connection like so:

In [12]:
!pip install -qU langchain-pinecone pinecone-notebooks

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.0/76.0 kB 5.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 587.6/587.6 kB 21.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 259.3/259.3 kB 23.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.5/65.5 kB 5.7 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pylibcudf-cu12 25.6.0 requires pyarrow<20.0.0a0,>=14.0.0; platform_machine == "x86_64", but you have pyarrow 21.0.0 which is incompatible.
cudf-cu12 25.6.0 requires pyarrow<20.0.0a0,>=14.0.0; platform_machine == "x86_64", but you have pyarrow 21.0.0 which is incompatible.


In [33]:
from pinecone import Pinecone, ServerlessSpec

spec = ServerlessSpec(
    cloud="aws", region="us-east-1"
)

# connect to pinecone environment
# connect to pinecone environment
pc = Pinecone(api_key=os.environ["PINECONE_API_KEY"])
print("✅ Pinecone initialized successfully")

# list indexes to confirm connection
print("Available indexes:", pc.list_indexes().names())

✅ Pinecone initialized successfully
Available indexes: ['extractive-question-answering']


In [34]:
from google.colab import userdata
import os

# Load keys from Colab Secrets into environment
os.environ["PINECONE_API_KEY"] = userdata.get("PINECONE_API_KEY")
os.environ["OPENAI_API_KEY"] = userdata.get("OPENAI_API_KEY")
os.environ["HF_TOKEN"] = userdata.get("HF_TOKEN")

print("Pinecone key loaded:", bool(os.environ.get("PINECONE_API_KEY")))

Pinecone key loaded: True


Now we create a new index called "question-answering" — we can name the index anything we want. We specify the metric type as "cosine" and dimension as 384 because the retriever we use to generate context embeddings is optimized for cosine similarity and outputs 384-dimension vectors.

In [46]:
import time

index_name = "extractive-question-answering"

if index_name in pc.list_indexes().names():
    pc.delete_index(index_name)

    while index_name in pc.list_indexes().names():
        time.sleep(1)

pc.create_index(
    name=index_name,
    dimension=384,
    metric="cosine",
    spec=spec
)

# connect
index = pc.Index(index_name)
print(f"Recreated & connected to index: {index_name}")

Recreated & connected to index: extractive-question-answering


# Initialize Retriever

Next, we need to initialize our retriever. The retriever will mainly do two things:

- Generate embeddings for all context passages (context vectors/embeddings)
- Generate embeddings for our questions (query vector/embedding)

The retriever will generate embeddings in a way that the questions and context passages containing answers to our questions are nearby in the vector space. We can use cosine similarity to calculate the similarity between the query and context embeddings to find the context passages that contain potential answers to our question.

We will use a SentenceTransformer model named ``multi-qa-MiniLM-L6-cos-v1`` designed for semantic search and trained on 215M (question, answer) pairs from diverse sources as our retriever.

In [47]:
import torch
from sentence_transformers import SentenceTransformer

# set device to GPU if available
device = 'cuda' if torch.cuda.is_available() else 'cpu'
# load the retriever model from huggingface model hub
retriever = SentenceTransformer("multi-qa-MiniLM-L6-cos-v1", device=device)
retriever

SentenceTransformer(
  (0): Transformer({'max_seq_length': 512, 'do_lower_case': False, 'architecture': 'BertModel'})
  (1): Pooling({'word_embedding_dimension': 384, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
)

# Generate Embeddings and Upsert

Next, we need to generate embeddings for the context passages. We will do this in batches to help us more quickly generate embeddings and upload them to the Pinecone index. When passing the documents to Pinecone, we need an id (a unique value), context embedding, and metadata for each document representing context passages in the dataset. The metadata is a dictionary containing data relevant to our embeddings, such as the article title, context passage, etc.

In [48]:
from tqdm.auto import tqdm

# we will use batches of 64
batch_size = 64

for i in tqdm(range(0, len(df), batch_size)):
    # find end of batch
    end = min(i + batch_size, len(df))
    # extract batch
    batch = df.iloc[i:end]
    # generate embeddings for batch
    emb = retriever.encode(batch["context"].tolist(), normalize_embeddings=True)
    # get metadata
    meta = [{"title": t, "context": c} for t, c in zip(batch["title"], batch["context"])]
    # create unique IDs
    ids = [f"squad-{k}" for k in range(i, end)]
    # add all to upsert list
    to_upsert = [
        {"id": _id, "values": vec.tolist(), "metadata": m}
        for _id, vec, m in zip(ids, emb, meta)
    ]
    # upsert/insert these records to pinecone
    _ = index.upsert(vectors=to_upsert)

index.describe_index_stats()

  0%|          | 0/296 [00:00<?, ?it/s]

{'dimension': 384,
 'index_fullness': 0.0,
 'metric': 'cosine',
 'namespaces': {'': {'vector_count': 18816}},
 'total_vector_count': 18816,
 'vector_type': 'dense'}

# Initialize Reader

We use the `deepset/electra-base-squad2` model from the HuggingFace model hub as our reader model. We load this model into a "question-answering" pipeline from HuggingFace transformers and feed it our questions and context passages individually. The model gives a prediction for each context we pass through the pipeline.

In [49]:
from transformers import pipeline

model_name = 'deepset/electra-base-squad2'
# load the reader model into a question-answering pipeline
reader = pipeline(tokenizer=model_name, model=model_name, task='question-answering', device=device)
reader

config.json:   0%|          | 0.00/635 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/436M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/200 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

Device set to use cuda


Now all the components we need are ready. Let's write some helper functions to execute our queries. The `get_context` function retrieves the context embeddings containing answers to our question from the Pinecone index, and the `extract_answer` function extracts the answers from these context passages.

In [50]:
# gets context passages from the pinecone index
def get_context(question, top_k):
    # generate embeddings for the question
    xq = retriever.encode(question).tolist()

    # search pinecone index for context passages
    xc = index.query(vector=xq, top_k=top_k, include_metadata=True)

    # extract the context passages from pinecone search results
    c = [match['metadata']['context'] for match in xc['matches']]

    return c

In [51]:
from pprint import pprint

# extracts answer from the context passage
def extract_answer(question, context):
    results = []
    for c in context:
        # feed the reader the question and contexts to extract answers
        answer = reader(question=question, context=c)
        # add the context to answer dict for printing both together
        answer["context"] = c
        results.append(answer)
    # sort the result based on the score from reader model
    sorted_result = pprint(sorted(results, key=lambda x: x['score'], reverse=True))
    return sorted_result

In [52]:
question = "How much oil is Egypt producing in a day?"
context = get_context(question, top_k = 1)
context

['Egypt was producing 691,000 bbl/d of oil and 2,141.05 Tcf of natural gas (in 2013), which makes Egypt as the largest oil producer not member of the Organization of the Petroleum Exporting Countries (OPEC) and the second-largest dry natural gas producer in Africa. In 2013, Egypt was the largest consumer of oil and natural gas in Africa, as more than 20% of total oil consumption and more than 40% of total dry natural gas consumption in Africa. Also, Egypt possesses the largest oil refinery capacity in Africa 726,000 bbl/d (in 2012). Egypt is currently planning to build its first nuclear power plant in El Dabaa city, northern Egypt.']

As we can see, the retiever is working fine and gets us the context passage that contains the answer to our question. Now let's use the reader to extract the exact answer from the context passage.

In [53]:
extract_answer(question, context)

[{'answer': '691,000 bbl/d',
  'context': 'Egypt was producing 691,000 bbl/d of oil and 2,141.05 Tcf of '
             'natural gas (in 2013), which makes Egypt as the largest oil '
             'producer not member of the Organization of the Petroleum '
             'Exporting Countries (OPEC) and the second-largest dry natural '
             'gas producer in Africa. In 2013, Egypt was the largest consumer '
             'of oil and natural gas in Africa, as more than 20% of total oil '
             'consumption and more than 40% of total dry natural gas '
             'consumption in Africa. Also, Egypt possesses the largest oil '
             'refinery capacity in Africa 726,000 bbl/d (in 2012). Egypt is '
             'currently planning to build its first nuclear power plant in El '
             'Dabaa city, northern Egypt.',
  'end': 33,
  'score': 0.9999855201993455,
  'start': 20}]


The reader model predicted with 99% accuracy the correct answer *691,000 bbl/d* as seen from the context passage. Let's run few more queries.

In [54]:
question = "What are the first names of the men that invented youtube?"
context = get_context(question, top_k=1)
extract_answer(question, context)

[{'answer': 'Hurley and Chen',
  'context': 'According to a story that has often been repeated in the media, '
             'Hurley and Chen developed the idea for YouTube during the early '
             'months of 2005, after they had experienced difficulty sharing '
             "videos that had been shot at a dinner party at Chen's apartment "
             'in San Francisco. Karim did not attend the party and denied that '
             'it had occurred, but Chen commented that the idea that YouTube '
             'was founded after a dinner party "was probably very strengthened '
             'by marketing ideas around creating a story that was very '
             'digestible".',
  'end': 79,
  'score': 0.9999276399612427,
  'start': 64}]


In [55]:
question = "What is Albert Eistein famous for?"
context = get_context(question, top_k=1)
extract_answer(question, context)

[{'answer': 'his theories of special relativity and general relativity',
  'context': 'Albert Einstein is known for his theories of special relativity '
             'and general relativity. He also made important contributions to '
             'statistical mechanics, especially his mathematical treatment of '
             'Brownian motion, his resolution of the paradox of specific '
             'heats, and his connection of fluctuations and dissipation. '
             'Despite his reservations about its interpretation, Einstein also '
             'made contributions to quantum mechanics and, indirectly, quantum '
             'field theory, primarily through his theoretical studies of the '
             'photon.',
  'end': 86,
  'score': 0.9500371217727661,
  'start': 29}]


Let's run another question. This time for top 3 context passages from the retriever.

In [56]:
question = "Who was the first person to step foot on the moon?"
context = get_context(question, top_k=3)
extract_answer(question, context)

[{'answer': 'Armstrong',
  'context': 'The trip to the Moon took just over three days. After achieving '
             'orbit, Armstrong and Aldrin transferred into the Lunar Module, '
             'named Eagle, and after a landing gear inspection by Collins '
             'remaining in the Command/Service Module Columbia, began their '
             'descent. After overcoming several computer overload alarms '
             'caused by an antenna switch left in the wrong position, and a '
             'slight downrange error, Armstrong took over manual flight '
             'control at about 180 meters (590 ft), and guided the Lunar '
             'Module to a safe landing spot at 20:18:04 UTC, July 20, 1969 '
             '(3:17:04 pm CDT). The first humans on the Moon would wait '
             'another six hours before they ventured out of their craft. At '
             '02:56 UTC, July 21 (9:56 pm CDT July 20), Armstrong became the '
             'first human to set foot on the Moon.',

The result looks pretty good.

In [57]:
pc.delete_index(index_name)

### Add a few more questions. What did you observe?

In [78]:
from pinecone import Pinecone, ServerlessSpec
pc = Pinecone(api_key=os.environ["PINECONE_API_KEY"])
spec = ServerlessSpec(cloud="aws", region="us-east-1")

In [79]:
index_name = "extractive-question-answering"

if index_name not in pc.list_indexes().names():
    pc.create_index(
        name=index_name,
        dimension=384,
        metric="cosine",
        spec=spec
    )

index = pc.Index(index_name)
index.describe_index_stats()

{'dimension': 384,
 'index_fullness': 0.0,
 'metric': 'cosine',
 'namespaces': {'': {'vector_count': 18891}},
 'total_vector_count': 18891,
 'vector_type': 'dense'}

In [80]:
import torch
from sentence_transformers import SentenceTransformer

device = 'cuda' if torch.cuda.is_available() else 'cpu'
retriever = SentenceTransformer("multi-qa-MiniLM-L6-cos-v1", device=device)

In [81]:
from tqdm.auto import tqdm

batch_size = 64

for i in tqdm(range(0, len(df), batch_size)):
    end   = min(i + batch_size, len(df))
    batch = df.iloc[i:end]

    emb  = retriever.encode(batch["context"].tolist(), normalize_embeddings=True)
    meta = [{"title": t, "context": c} for t, c in zip(batch["title"], batch["context"])]
    ids  = [f"squad-{k}" for k in range(i, end)]

    to_upsert = [{"id": _id, "values": vec.tolist(), "metadata": m}
                 for _id, vec, m in zip(ids, emb, meta)]

    _ = index.upsert(vectors=to_upsert)

index.describe_index_stats()

  0%|          | 0/296 [00:00<?, ?it/s]

{'dimension': 384,
 'index_fullness': 0.0,
 'metric': 'cosine',
 'namespaces': {'': {'vector_count': 18891}},
 'total_vector_count': 18891,
 'vector_type': 'dense'}

In [85]:
from pprint import pprint

def extract_answer(question, context):
    results = []
    for c in context:
        ans = reader(question=question, context=c)
        ans["context"] = c
        results.append(ans)

    results = sorted(results, key=lambda x: x["score"], reverse=True)

    best = results[0]
    pprint(best)

In [83]:
from transformers import pipeline
reader = pipeline(tokenizer='deepset/electra-base-squad2',
                  model='deepset/electra-base-squad2',
                  task='question-answering')

from pprint import pprint

def extract_answer(question, context):
    results = []
    for c in context:
        answer = reader(question=question, context=c)
        answer["context"] = c
        results.append(answer)
    sorted_result = pprint(sorted(results, key=lambda x: x['score'], reverse=True))
    return sorted_result

Device set to use cuda:0


### Question 1

In [86]:
question = "Who discovered penicillin?"
context  = get_context(question, top_k=3)
extract_answer(question, context)

{'answer': 'Alexander Fleming',
 'context': 'The effects of some types of mold on infection had been noticed '
            'many times over the course of history (see: History of '
            'penicillin). In 1928, Alexander Fleming noticed the same effect '
            'in a Petri dish, where a number of disease-causing bacteria were '
            'killed by a fungus of the genus Penicillium. Fleming postulated '
            'that the effect is mediated by an antibacterial compound he named '
            'penicillin, and that its antibacterial properties could be '
            'exploited for chemotherapy. He initially characterized some of '
            'its biological properties, and attempted to use a crude '
            'preparation to treat some infections, but he was unable to pursue '
            'its further development without the aid of trained chemists.',
 'end': 158,
 'score': 0.6873158812522888,
 'start': 141}


In [87]:
question = "Who developed the theory of relativity?"
context  = get_context(question, top_k=3)
extract_answer(question, context)

{'answer': 'Einstein',
 'context': 'The beginning of the 20th century brought the start of a '
            'revolution in physics. The long-held theories of Newton were '
            'shown not to be correct in all circumstances. Beginning in 1900, '
            'Max Planck, Albert Einstein, Niels Bohr and others developed '
            'quantum theories to explain various anomalous experimental '
            'results, by introducing discrete energy levels. Not only did '
            'quantum mechanics show that the laws of motion did not hold on '
            'small scales, but even more disturbingly, the theory of general '
            'relativity, proposed by Einstein in 1915, showed that the fixed '
            'background of spacetime, on which both Newtonian mechanics and '
            'special relativity depended, could not exist. In 1925, Werner '
            'Heisenberg and Erwin Schrödinger formulated quantum mechanics, '
            'which explained the preceding quantum the

### Example 3

In [88]:
question = "What is the capital of France?"
context  = get_context(question, top_k=3)
extract_answer(question, context)

You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset


{'answer': 'London',
 'context': 'Paris is located in northern central France. By road it is 450 '
            'kilometres (280 mi) south-east of London, 287 kilometres (178 mi) '
            'south of Calais, 305 kilometres (190 mi) south-west of Brussels, '
            '774 kilometres (481 mi) north of Marseille, 385 kilometres (239 '
            'mi) north-east of Nantes, and 135 kilometres (84 mi) south-east '
            'of Rouen. Paris is located in the north-bending arc of the river '
            'Seine and includes two islands, the Île Saint-Louis and the '
            'larger Île de la Cité, which form the oldest part of the city. '
            "The river's mouth on the English Channel (La Manche) is about 233 "
            'mi (375 km) downstream of the city, established around 7600 BC. '
            'The city is spread widely on both banks of the river. Overall, '
            'the city is relatively flat, and the lowest point is 35 m (115 '
            'ft) above sea level.

## Observation

#### Example 1: “Who discovered penicillin?”
- **Predicted Answer:** Alexander Fleming  
- **Confidence Score:** 0.6873  
- **Observation:**  
  The model correctly identified **Alexander Fleming** as the discoverer of penicillin.  
  The retrieved context was highly relevant, describing both the discovery and its historical background.  
  This shows the model’s ability to locate accurate, context-supported information for scientific questions.



#### Example 2: “Who developed the theory of relativity?”
- **Predicted Answer:** Einstein  
- **Confidence Score:** 0.9999  
- **Observation:**  
  The model produced a perfect and concise answer (**Einstein**) with a very high confidence score, showing strong contextual alignment.  
  The context clearly mentioned relativity and Einstein’s contribution, demonstrating that both the retriever and reader were working effectively together.



#### Example 3: “What is the capital of France?”
- **Predicted Answer:** London  
- **Confidence Score:** 1.48e-07  
- **Observation:**  
  The model returned an incorrect answer even though the retrieved passage correctly described **Paris**.  
  This error likely stems from the model’s inability to correctly extract short factual spans in some cases.  
  
  **Conclusion:**  
  The retriever fetched the correct passage, but the reader failed to pinpoint the exact answer.  
  This highlights how model size and language handling can affect extractive question-answering performance.



### Insight
The experiment demonstrates that while the retriever and reader pipeline generally performs well for scientific and historical questions,  
it can misfire on simpler factual ones depending on the model version and embedding precision.  
Using a larger, monolingual model (like **mpnet-base-dot-v1**) could improve factual accuracy at the cost of speed.